packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -2731,51 +2721,49 @@ export class ContainerRuntime
 				return;
 			}
 
+			let runtimeBatch: boolean = true;
 			// Reach out to PendingStateManager, either to zip localOpMetadata into the *local* message list,
 			// or to check to ensure the *remote* messages don't match the batchId of a pending local batch.
 			// This latter case would indicate that the container has forked - two copies are trying to persist the same local changes.
-			const messagesWithPendingState = this.pendingStateManager.processInboundBatch(
-				inboundBatch,
-				local,
-			);
-			if (messagesWithPendingState.length > 0) {
-				messagesWithPendingState.forEach(({ message, localOpMetadata }) => {
-					const msg: MessageWithContext = {
-						message,
-						local,
-						isRuntimeMessage: true,
-						savedOp,
-						localOpMetadata,
-					};
-					this.ensureNoDataModelChanges(() => this.processRuntimeMessage(msg));
-				});
-			} else {
-				this.ensureNoDataModelChanges(() => this.processEmptyBatch(inboundBatch));
+			let messagesWithPendingState: {
+				message: ISequencedDocumentMessage;
+				localOpMetadata?: unknown;
+			}[] = this.pendingStateManager.processInboundBatch(inboundBatch, local);
+			if (inboundBatch.messages.length === 0) {
+				/**
+				 * We need to process an empty batch, which will execute expected actions while processing even if there
+				 * are no inner runtime messages.
+				 *
+				 * Empty batches are produced by the outbox on resubmit when the resubmit flow resulted in no runtime
+				 * messages.
+				 * This can happen if changes from a remote client "cancel out" the pending changes being resubmitted by
+				 * this client.  We submit an empty batch if "offline load" (aka rehydrating from stashed state) is
+				 * enabled, to ensure we account for this batch when comparing batchIds, checking for a forked container.
+				 * Otherwise, we would not realize this container has forked in the case where it did fork, and a batch
+				 * became empty but wasn't submitted as such.
+				 */
+				const { keyMessage, batchStartCsn } = inboundBatch;
+				messagesWithPendingState = [
+					{
+						message: {
+							...keyMessage,
+							clientSequenceNumber: batchStartCsn,
```

Comment:
This is a no-op, try switching it to an assert.  See the `isGroupedBatch` case in `RemoteMessageProcessor.process`

Reply:
I see. I looked at the code and got confused. In case of an empty batch, the "message" there is the grouped message which has "messages" which is an empty array. So, in that case, it uses the grouped "message" as the key message.